File: src/pricing.py
"""Price, payoff, notional, margin and liquidation-cutoff math.

Implements spec sections 2-4. The key invariants enforced here:

* ``ScenarioPrice(z) = EntryPrice * (1 + z)`` (NEVER CurrentPrice/FinalPrice).
* Long payoff = ``z``; short payoff = ``-z``.
* PnL is sized by the *open notional* ``N`` only. Adding margin changes margin
  and liquidation cutoff, never ``N``.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional


def scenario_price(entry_price: float, z: float) -> float:
    """Final-scenario price for return ``z`` relative to the entry price."""
    return float(entry_price) * (1.0 + float(z))


def y_long(current_price: float, entry_price_long: float) -> float:
    return float(current_price) / float(entry_price_long) - 1.0


def y_short(current_price: float, entry_price_short: float) -> float:
    return float(current_price) / float(entry_price_short) - 1.0


def long_payoff_pct(z: float) -> float:
    return float(z)


def short_payoff_pct(z: float) -> float:
    return -float(z)


def notional(qty: float, entry_price: float) -> float:
    return float(qty) * float(entry_price)


def risk_leverage(notional_value: float, margin_current: float) -> float:
    if margin_current == 0:
        return float("inf")
    return float(notional_value) / float(margin_current)


def long_liq_z(
    entry_price_long: float,
    notional_long: float,
    margin_long_current: float,
    liquidation_price_long: Optional[float] = None,
) -> float:
    """Long liquidation cutoff in ``z`` units.

    Uses the real Binance liquidation price when available; otherwise the
    approximation ``-M_long_current / N_long``.
    """
    if liquidation_price_long is not None:
        return float(liquidation_price_long) / float(entry_price_long) - 1.0
    return -float(margin_long_current) / float(notional_long)


def short_liq_z(
    entry_price_short: float,
    notional_short: float,
    margin_short_current: float,
    liquidation_price_short: Optional[float] = None,
) -> float:
    """Short liquidation cutoff in ``z`` units.

    Real liquidation price when available; otherwise approximation
    ``M_short_current / N_short``.
    """
    if liquidation_price_short is not None:
        return float(liquidation_price_short) / float(entry_price_short) - 1.0
    return float(margin_short_current) / float(notional_short)


def liquidation_price_from_cutoff(entry_price: float, liq_z: float) -> float:
    """Inverse of the cutoff: price level that triggers liquidation."""
    return float(entry_price) * (1.0 + float(liq_z))


def add_margin(margin_current: float, x: float) -> float:
    """M_X = M_current + X. Notional is intentionally NOT touched here."""
    return float(margin_current) + float(x)


@dataclass
class LegState:
    """State of a single position leg. ``notional`` is fixed after open."""

    side: str  # "long" or "short"
    entry_price: float
    qty: float
    margin_current: float
    liquidation_price: Optional[float] = None

    @property
    def notional(self) -> float:
        return notional(self.qty, self.entry_price)

    def current_return(self, current_price: float) -> float:
        return float(current_price) / float(self.entry_price) - 1.0

    def payoff_pct(self, z: float) -> float:
        return long_payoff_pct(z) if self.side == "long" else short_payoff_pct(z)

    def liq_z(self) -> float:
        if self.side == "long":
            return long_liq_z(
                self.entry_price, self.notional, self.margin_current, self.liquidation_price
            )
        return short_liq_z(
            self.entry_price, self.notional, self.margin_current, self.liquidation_price
        )

    def liq_z_after_add(self, x: float) -> float:
        """Liquidation cutoff after adding margin ``x`` (notional unchanged)."""
        m_x = add_margin(self.margin_current, x)
        if self.side == "long":
            return long_liq_z(self.entry_price, self.notional, m_x, None)
        return short_liq_z(self.entry_price, self.notional, m_x, None)

    def current_pnl(self, current_price: float) -> float:
        """Realized-if-closed-now PnL = N_open * payoff_pct(y)."""
        y = self.current_return(current_price)
        return self.notional * self.payoff_pct(y)

    def liquidation_price_level(self) -> float:
        if self.liquidation_price is not None:
            return float(self.liquidation_price)
        return liquidation_price_from_cutoff(self.entry_price, self.liq_z())

File: rust_live/src/pricing.rs
//! Price / payoff / notional / liquidation-cutoff math.
//!
//! Invariants: `ScenarioPrice(z) = EntryPrice * (1 + z)`, long payoff = `z`,
//! short payoff = `-z`, and PnL is sized by the open notional `N` only. Adding
//! margin changes margin and the liquidation cutoff, never `N`.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum Side {
    Long,
    Short,
}

impl Side {
    pub fn as_str(&self) -> &'static str {
        match self {
            Side::Long => "long",
            Side::Short => "short",
        }
    }
    pub fn position_side(&self) -> &'static str {
        match self {
            Side::Long => "LONG",
            Side::Short => "SHORT",
        }
    }
}

pub fn scenario_price(entry_price: f64, z: f64) -> f64 {
    entry_price * (1.0 + z)
}

pub fn long_payoff_pct(z: f64) -> f64 {
    z
}

pub fn short_payoff_pct(z: f64) -> f64 {
    -z
}

pub fn notional(qty: f64, entry_price: f64) -> f64 {
    qty * entry_price
}

pub fn long_liq_z(entry_price: f64, n: f64, m: f64, liq_price: Option<f64>) -> f64 {
    match liq_price {
        Some(p) => p / entry_price - 1.0,
        None => -m / n,
    }
}

pub fn short_liq_z(entry_price: f64, n: f64, m: f64, liq_price: Option<f64>) -> f64 {
    match liq_price {
        Some(p) => p / entry_price - 1.0,
        None => m / n,
    }
}

pub fn liquidation_price_from_cutoff(entry_price: f64, liq_z: f64) -> f64 {
    entry_price * (1.0 + liq_z)
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LegState {
    pub side: Side,
    pub entry_price: f64,
    pub qty: f64,
    pub margin_current: f64,
    #[serde(default)]
    pub liquidation_price: Option<f64>,
}

impl LegState {
    pub fn new(side: Side, entry_price: f64, qty: f64, margin_current: f64) -> Self {
        Self {
            side,
            entry_price,
            qty,
            margin_current,
            liquidation_price: None,
        }
    }

    pub fn notional(&self) -> f64 {
        notional(self.qty, self.entry_price)
    }

    pub fn current_return(&self, current_price: f64) -> f64 {
        current_price / self.entry_price - 1.0
    }

    pub fn payoff_pct(&self, z: f64) -> f64 {
        match self.side {
            Side::Long => long_payoff_pct(z),
            Side::Short => short_payoff_pct(z),
        }
    }

    pub fn liq_z(&self) -> f64 {
        match self.side {
            Side::Long => long_liq_z(self.entry_price, self.notional(), self.margin_current, self.liquidation_price),
            Side::Short => short_liq_z(self.entry_price, self.notional(), self.margin_current, self.liquidation_price),
        }
    }

    /// Liquidation cutoff after adding margin `x` (notional unchanged).
    pub fn liq_z_after_add(&self, x: f64) -> f64 {
        let m_x = self.margin_current + x;
        match self.side {
            Side::Long => long_liq_z(self.entry_price, self.notional(), m_x, None),
            Side::Short => short_liq_z(self.entry_price, self.notional(), m_x, None),
        }
    }

    pub fn current_pnl(&self, current_price: f64) -> f64 {
        let y = self.current_return(current_price);
        self.notional() * self.payoff_pct(y)
    }

    pub fn liquidation_price_level(&self) -> f64 {
        match self.liquidation_price {
            Some(p) => p,
            None => liquidation_price_from_cutoff(self.entry_price, self.liq_z()),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn scenario_uses_entry() {
        assert!((scenario_price(100.0, 0.1) - 110.0).abs() < 1e-9);
    }

    #[test]
    fn payoff_signs() {
        assert_eq!(long_payoff_pct(0.2), 0.2);
        assert_eq!(short_payoff_pct(0.2), -0.2);
    }

    #[test]
    fn add_margin_keeps_notional() {
        let mut leg = LegState::new(Side::Long, 100.0, 10.0, 100.0);
        let n_before = leg.notional();
        leg.margin_current += 50.0;
        assert!((leg.notional() - n_before).abs() < 1e-12);
    }

    #[test]
    fn liq_z_approximations() {
        assert!((long_liq_z(100.0, 1000.0, 100.0, None) + 0.1).abs() < 1e-12);
        assert!((short_liq_z(100.0, 1000.0, 100.0, None) - 0.1).abs() < 1e-12);
    }
}
